🐛 BUG: asset bundling resolution ENOENT error using TypeScript path aliases

### Which Cloudflare product(s) does this pertain to?

Wrangler

### What version of `Wrangler` are you using?

2.13.0

### What operating system are you using?

Mac

### Describe the Bug

Getting the following module resolution error when resolving a non-standard extension (`.graphql`) while using TS path aliases.

**Please note:** this works fine if I don't use the `~/*` alias I have configured and use a relative import instead. Just the alias seems to mess up the file resolution.

```
[ERROR] [plugin wrangler-module-collector] ENOENT: no such file or directory, open '/Users/estrattonbailey/Sites/rainbow/arc/workers/nfts/src/routers/graphql/~/routers/graphql/schema.graphql'

    src/routers/graphql/index.ts:7:19:
      7 │ import schema from '~/routers/graphql/schema.graphql';
        ╵                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  This error came from the "onResolve" callback registered here:

    ../../node_modules/.pnpm/wrangler@2.13.0/node_modules/wrangler/wrangler-dist/cli.js:130799:19:
      130799 │             build5.onResolve(
             ╵                    ~~~~~~~~~

    at /Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/wrangler@2.13.0/node_modules/wrangler/wrangler-dist/cli.js:130799:20
    at Array.forEach (<anonymous>)
    at /Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/wrangler@2.13.0/node_modules/wrangler/wrangler-dist/cli.js:130798:22
    at Array.forEach (<anonymous>)
    at setup (/Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/wrangler@2.13.0/node_modules/wrangler/wrangler-dist/cli.js:130795:16)
    at handlePlugins (/Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/esbuild@0.16.3/node_modules/esbuild/lib/main.js:1276:21)
    at buildOrServeImpl (/Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/esbuild@0.16.3/node_modules/esbuild/lib/main.js:965:5)
    at Object.buildOrServe (/Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/esbuild@0.16.3/node_modules/esbuild/lib/main.js:773:5)
    at /Users/estrattonbailey/Sites/rainbow/arc/node_modules/.pnpm/esbuild@0.16.3/node_modules/esbuild/lib/main.js:2112:17
```

I'm importing the following:

```typescript
import schema from '~/routers/graphql/schema.graphql';
...
```

Here's my `wrangler.toml`:

```toml
rules = [
  { type = "Text", globs = ["**/*.graphql"], fallthrough = true }
]
```

And here's my `tsconfig.json`:
```jsonc
{
  "include": ["src", "env.d.ts"],
  "compilerOptions": {
    "strict": true /* Enable all strict type-checking options. */,
    "target": "es2021" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
    "lib": [
      "es2021"
    ] /* Specify a set of bundled library declaration files that describe the target runtime environment. */,
    "module": "es2022" /* Specify what module code is generated. */,
    "moduleResolution": "node" /* Specify how TypeScript looks up a file from a given module specifier. */,
    "baseUrl": "./" /* Specify the base directory to resolve non-relative module names. */,
    "paths": {
      "~/*": ["src/*"]
    } /* Specify a set of entries that re-map imports to additional lookup locations. */,
    "types": [
      "@cloudflare/workers-types"
    ] /* Specify type package names to be included without being referenced in a source file. */,
    "resolveJsonModule": true /* Enable importing .json files */,
    "allowJs": true /* Allow JavaScript files to be a part of your program. Use the `checkJS` option to get errors from these files. */,
    "checkJs": false /* Enable error reporting in type-checked JavaScript files. */,
    "noEmit": true /* Disable emitting files from a compilation. */,
    "isolatedModules": true /* Ensure that each file can be safely transpiled without relying on other imports. */,
    "esModuleInterop": true /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables `allowSyntheticDefaultImports` for type compatibility. */,
    "forceConsistentCasingInFileNames": true /* Ensure that casing is correct in imports. */,
    // "skipDefaultLibCheck": true, /* Skip type checking .d.ts files that are included with TypeScript. */
    "skipLibCheck": true /* Skip type checking all .d.ts files. */
  }
}
```

--- a/.changeset/fix-tsconfig-paths-module-collection.md
+++ b/.changeset/fix-tsconfig-paths-module-collection.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+fix: respect TypeScript path aliases when resolving non-JS modules with module rules
+
+When importing non-JavaScript files (like `.graphql`, `.txt`, etc.) using TypeScript path aliases defined in `tsconfig.json`, Wrangler's module-collection plugin now correctly resolves these imports. Previously, path aliases were only respected for JavaScript/TypeScript files, causing imports like `import schema from '~lib/schema.graphql'` to fail when using module rules.

--- a/packages/wrangler/src/__tests__/deploy.test.ts
+++ b/packages/wrangler/src/__tests__/deploy.test.ts
@@ -11884,6 +11884,39 @@ addEventListener('fetch', event => {});`
 			`);
 		});
 
+		it("should use compilerOptions.paths to resolve non-js modules with module rules", async () => {
+			writeWranglerConfig({
+				main: "index.ts",
+				rules: [{ type: "Text", globs: ["**/*.graphql"], fallthrough: true }],
+			});
+			fs.writeFileSync(
+				"index.ts",
+				`import schema from '~lib/schema.graphql'; export default { fetch() { return new Response(schema)} }`
+			);
+			fs.mkdirSync("lib", { recursive: true });
+			fs.writeFileSync("lib/schema.graphql", `type Query { hello: String }`);
+			fs.writeFileSync(
+				"tsconfig.json",
+				JSON.stringify({
+					compilerOptions: {
+						baseUrl: ".",
+						paths: {
+							"~lib/*": ["lib/*"],
+						},
+					},
+				})
+			);
+			mockSubDomainRequest();
+			mockUploadWorkerRequest({
+				expectedModules: {
+					"./bc4a21e10be4cae586632dfe5c3f049299c06466-schema.graphql":
+						"type Query { hello: String }",
+				},
+			});
+			await runWrangler("deploy index.ts");
+			expect(std.err).toMatchInlineSnapshot(`""`);
+		});
+
 		it("should output to target es2022 even if tsconfig says otherwise", async () => {
 			writeWranglerConfig();
 			writeWorkerSource();

--- a/packages/wrangler/src/deployment-bundle/module-collection.ts
+++ b/packages/wrangler/src/deployment-bundle/module-collection.ts
@@ -278,7 +278,8 @@ export function createModuleCollector(props: {
 								// and resolve the file path to the correct file.
 								try {
 									const resolved = await build.resolve(args.path, {
-										kind: "import-statement",
+										kind: args.kind,
+										importer: args.importer,
 										resolveDir: args.resolveDir,
 										pluginData: {
 											skip: true,

